coderd/cryptokeys/dbkeycache.go

```diff
@@ -0,0 +1,186 @@
+	keysMu    sync.RWMutex
+	keys      map[int32]database.CryptoKey
+	latestKey database.CryptoKey
+	fetched   chan struct{}
+}
+
+type DBCacheOption func(*DBCache)
+
+func WithDBCacheClock(clock quartz.Clock) DBCacheOption {
+	return func(d *DBCache) {
+		d.clock = clock
+	}
+}
+
+// NewDBCache creates a new DBCache. It starts a background
+// process that periodically refreshes the cache. The context should
+// be canceled to stop the background process.
+func NewDBCache(ctx context.Context, logger slog.Logger, db database.Store, feature database.CryptoKeyFeature, opts ...func(*DBCache)) *DBCache {
+	d := &DBCache{
+		db:      db,
+		feature: feature,
+		clock:   quartz.NewReal(),
+		logger:  logger,
+		fetched: make(chan struct{}),
+	}
+
+	for _, opt := range opts {
+		opt(d)
+	}
+
+	go d.clear(ctx)
+	return d
+}
+
+// Verifying returns the CryptoKey with the given sequence number, provided that
+// it is neither deleted nor has breached its deletion date. It should only be
+// used for verifying or decrypting payloads. To sign/encrypt call Signing.
+func (d *DBCache) Verifying(ctx context.Context, sequence int32) (codersdk.CryptoKey, error) {
+	now := d.clock.Now()
+	d.keysMu.RLock()
+	key, ok := d.keys[sequence]
+	d.keysMu.RUnlock()
+	if ok {
+		return checkKey(key, now)
+	}
+
+	d.keysMu.Lock()
+	defer d.keysMu.Unlock()
+
+	key, ok = d.keys[sequence]
+	if ok {
+		return checkKey(key, now)
+	}
+
+	cache, latest, err := d.fetch(ctx)
+	if err != nil {
+		return codersdk.CryptoKey{}, xerrors.Errorf("fetch: %w", err)
+	}
+	d.keys, d.latestKey = cache, latest
+
+	key, ok = d.keys[sequence]
+	if !ok {
+		return codersdk.CryptoKey{}, ErrKeyNotFound
+	}
+
+	return checkKey(key, now)
+}
+
+// Signing returns the latest valid key for signing. A valid key is one that is
+// both past its start time and before its deletion time.
+func (d *DBCache) Signing(ctx context.Context) (codersdk.CryptoKey, error) {
+	d.keysMu.RLock()
+	latest := d.latestKey
+	d.keysMu.RUnlock()
+
+	now := d.clock.Now()
+	if latest.CanSign(now) {
+		return db2sdk.CryptoKey(latest), nil
+	}
+
+	d.keysMu.Lock()
+	defer d.keysMu.Unlock()
+
+	if d.latestKey.CanSign(now) {
+		return db2sdk.CryptoKey(d.latestKey), nil
+	}
+
+	// Refetch all keys for this feature so we can find the latest valid key.
+	cache, latest, err := d.fetch(ctx)
+	if err != nil {
+		return codersdk.CryptoKey{}, xerrors.Errorf("fetch: %w", err)
+	}
+	d.keys, d.latestKey = cache, latest
+
+	return db2sdk.CryptoKey(d.latestKey), nil
+}
+
+func (d *DBCache) clear(ctx context.Context) {
+	for {
+		fired := make(chan struct{})
+		timer := d.clock.AfterFunc(time.Minute*10, func() {
+			defer close(fired)
+
+			// There's a small window where the timer fires as we're fetching
+			// keys that could result in us immediately invalidating the cache that we just populated.
+			d.keysMu.Lock()
+			defer d.keysMu.Unlock()
+			d.keys = nil
+			d.latestKey = database.CryptoKey{}
+		})
+
+		select {
+		case <-ctx.Done():
```

Comment:
In order to not leak a goroutine in the timer, you need to stop it when the context completes (or add an explicit `Close()` method to the cache).

Should probably add `goleak` to the test suite.